File: src/main/java/backend/vagas/controller/exception/HandlerAdviceException.java
package backend.vagas.controller.exception;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.HttpRequestMethodNotSupportedException;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

import java.util.HashMap;
import java.util.Map;


@ControllerAdvice
public class HandlerAdviceException {


    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<Map<String, String>> methodArgumentNotValidException(MethodArgumentNotValidException e) {
        Map<String, String> errors = new HashMap<>();
        e.getBindingResult().getFieldErrors().forEach(error -> errors.put(error.getDefaultMessage(), error.getField()));
        return new ResponseEntity<>(errors, HttpStatus.BAD_REQUEST);
    }

    @ExceptionHandler(HttpRequestMethodNotSupportedException.class)
    public ResponseEntity<Map<String, String>> methodArgumentNotValidException(HttpRequestMethodNotSupportedException e) {
        Map<String, String> errors = new HashMap<>();
        errors.put("ERROR! ", e.getLocalizedMessage());

        return new ResponseEntity<>(errors, HttpStatus.BAD_REQUEST);
    }

    @ExceptionHandler(UserNotFoundException.class)
    public ResponseEntity<Map<String, String>> userNotFoundException(UserNotFoundException e) {
        Map<String, String> errors = new HashMap<>();
        errors.put("ERROR! ", e.getLocalizedMessage());

        return new ResponseEntity<>(errors, HttpStatus.NOT_FOUND);
    }
}

File: src/main/java/backend/vagas/controller/exception/HandlerAdviceException.kt
package backend.vagas.controller.exception

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.validation.FieldError
import org.springframework.web.HttpRequestMethodNotSupportedException
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import java.util.function.Consumer

@ControllerAdvice
class HandlerAdviceException {
    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun methodArgumentNotValidException(e: MethodArgumentNotValidException): ResponseEntity<Map<String?, String>> {
        val errors: MutableMap<String?, String> = HashMap()
        e.bindingResult.fieldErrors.forEach(Consumer { error: FieldError ->
            errors[error.defaultMessage] = error.field
        })
        return ResponseEntity(errors, HttpStatus.BAD_REQUEST)
    }

    @ExceptionHandler(HttpRequestMethodNotSupportedException::class)
    fun methodArgumentNotValidException(e: HttpRequestMethodNotSupportedException): ResponseEntity<Map<String, String>> {
        val errors: MutableMap<String, String> = HashMap()
        errors["ERROR! "] = e.localizedMessage

        return ResponseEntity(errors, HttpStatus.BAD_REQUEST)
    }

    @ExceptionHandler(UserException::class)
    fun userNotFoundException(e: UserException): ResponseEntity<Map<String, String>> {
        val errors: MutableMap<String, String> = HashMap()
        errors["ERROR! "] = e.localizedMessage

        return ResponseEntity(errors, HttpStatus.CONFLICT)
    }

    @ExceptionHandler(UsernameNotFoundException::class)
    fun usernameNotFoundException(e: UsernameNotFoundException): ResponseEntity<Map<String, String>> {
        val errors: MutableMap<String, String> = HashMap()
        errors["ERROR! "] = e.localizedMessage

        return ResponseEntity(errors, HttpStatus.UNAUTHORIZED)
    }

}

